msal/src/main/java/com/microsoft/identity/client/helper/BrokerHelperActivity.java

```diff
@@ -0,0 +1,81 @@
+
+import android.app.Activity;
+import android.content.Context;
+import android.content.Intent;
+import android.os.Bundle;
+import android.util.Log;
+import android.widget.TextView;
+
+import androidx.annotation.Nullable;
+
+import com.microsoft.identity.common.adal.internal.AuthenticationConstants;
+import com.microsoft.identity.common.internal.broker.BrokerValidator;
+import com.microsoft.identity.common.internal.broker.PackageHelper;
+import com.microsoft.identity.msal.R;
+
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+
+public class BrokerHelperActivity extends Activity {
+
+    TextView mPackageName;
+    TextView mSignature;
+    TextView mRedirect;
+    TextView mManifest;
+
+    String mManifestTemplate= "<activity android:name=\"com.microsoft.identity.client.BrowserTabActivity\">\n" +
+            "    <intent-filter>\n" +
+            "        <action android:name=\"android.intent.action.VIEW\" />\n" +
+            "        <category android:name=\"android.intent.category.DEFAULT\" />\n" +
+            "        <category android:name=\"android.intent.category.BROWSABLE\" />\n" +
+            "        <data\n" +
+            "            android:host=\"%s\"\n" +
+            "            android:path=\"%s\"\n" +
+            "            android:scheme=\"msauth\" />\n" +
+            "    </intent-filter>\n" +
+            "</activity>";
+
+    public static Intent createStartIntent(final Context context) {
+        final Intent intent = new Intent(context, BrokerHelperActivity.class);
+          return intent;
+    }
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.broker_helper);
+
+        mPackageName = findViewById(R.id.txtPackageName);
+        mSignature = findViewById(R.id.txtSignature);
+        mRedirect = findViewById(R.id.txtRedirect);
+        mManifest = findViewById(R.id.txtManifest);
+
+        mSignature.append(getSignature(false));
+        mRedirect.append(BrokerValidator.getBrokerRedirectUri(this.getApplicationContext(), this.getPackageName()));
+        mManifest.append(String.format(mManifestTemplate, this.getPackageName(), getSignature(false)));
+        mPackageName.append(this.getPackageName());
+
+
+    }
+
+    private String getSignature(Boolean urlEncode){
+
+        final PackageHelper info = new PackageHelper(this.getApplicationContext().getPackageManager());
+        final String signatureDigest = info.getCurrentSignatureForPackage(this.getPackageName());
+        String signature = "";
+
+        try {
+            if(urlEncode) {
+                signature = URLEncoder.encode(signatureDigest, AuthenticationConstants.ENCODING_UTF8);
+            }else{
+                signature = signatureDigest;
+            }
+        } catch (UnsupportedEncodingException e) {
+            Log.e("getSignature", "Error encoding signature digest", e);
+        }

```

Comment:
Do we also want to display an error on the `BrokerHelperActivity` when this happens? Looks like an empty string will be shown

Reply:
Hadn't considered that since I don't expect this to happen... wondering in this case if I should convert this to a runtime exception.

Reply:
If you sneaky it out it'll land in the same place. ;-)

Reply:
I chose to convert checked exception to runtime in this case.  So I'll log and then throw a runtime exception.